ldap/servers/slapd/slapi-memberof.c

```diff
@@ -0,0 +1,1280 @@
+static void report_error_msg(Slapi_MemberOfConfig *config, char* msg);
+
+void
+Xdump_cache_entry(Xmemberof_cached_value *double_check, const char *msg)
+{
+    int i;
+    for (i = 0; double_check[i].valid; i++) {
+        slapi_log_err(SLAPI_LOG_ERR, "linked_attributes", "dump_cache_entry: %s -> %s/%s\n",
+                      msg ? msg : "<no key>",
+                      double_check[i].group_dn_val ? double_check[i].group_dn_val : "NULL",
+                      double_check[i].nsuniqueid_val ? double_check[i].nsuniqueid_val : "NULL");
+    }
+}
+static Xmemberof_cached_value *
+Xancestors_cache_lookup(Slapi_MemberOfConfig *config, const char *ndn)
+{
+    Xmemberof_cached_value *e;
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
+    long int start;
+    struct timespec tsnow;
+#endif
+
+    Xcache_stat.total_lookup++;
+
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
+    if (clock_gettime(CLOCK_REALTIME, &tsnow) != 0) {
+        start = 0;
+    } else {
+        start = tsnow.tv_nsec;
+    }
+#endif
+
+    e = (Xmemberof_cached_value *) PL_HashTableLookupConst(config->ancestors_cache, (const void *) ndn);
+
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
+    if (start) {
+        if (clock_gettime(CLOCK_REALTIME, &tsnow) == 0) {
+            Xcache_stat.cumul_duration_lookup += (tsnow.tv_nsec - start);
+        }
+    }
+#endif
+
+    if (e)
+        Xcache_stat.successfull_lookup++;
+    return e;
+}
+
+/* allocates the plugin hashtable
+ * This hash table is used by operation and is protected from
+ * concurrent operations with the memberof_lock (if not usetxn, memberof_lock
+ * is not implemented and the hash table will be not used.
+ *
+ * The hash table contains all the DN of the entries for which the memberof
+ * attribute has been computed/updated during the current operation
+ *
+ * hash table should be empty at the beginning and end of the plugin callback
+ */
+
+void Xancestor_hashtable_entry_free(Xmemberof_cached_value *entry)
+{
+    int i;
+
+    for (i = 0; entry[i].valid; i++) {
+        slapi_ch_free((void **) &entry[i].group_dn_val);
+        slapi_ch_free((void **) &entry[i].group_ndn_val);
+        slapi_ch_free((void **) &entry[i].nsuniqueid_val);
+    }
+    /* Here we are at the ending element containing the key */
+    slapi_ch_free((void**) &entry[i].key);
+}
+
+static PLHashEntry *
+Xancestors_cache_add(Slapi_MemberOfConfig *config, const void *key, void *value)
+{
+    PLHashEntry *e;
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
+    long int start;
+    struct timespec tsnow;
+#endif
+    Xcache_stat.total_add++;
+
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
+    if (clock_gettime(CLOCK_REALTIME, &tsnow) != 0) {
+        start = 0;
+    } else {
+        start = tsnow.tv_nsec;
+    }
+#endif
+
+    e = PL_HashTableAdd(config->ancestors_cache, key, value);
+
+#if defined(DEBUG) && defined(HAVE_CLOCK_GETTIME)
```

Comment:
@Firstyear  @tbordaz Off topic here, but do we _still_ need to check for HAVE_CLOCK_GETTIME?  I actually wanted to remove these define checks in a previous PR but didn't.  It would be nice (in a different PR of course) to get rid of them if we don't need them anymore.